firefox-ios/Client/Frontend/Browser/Toolbars/Models/AddressToolbarContainerModel.swift

```diff
@@ -71,10 +74,24 @@ class AddressToolbarContainerModel: Equatable {
         return query
     }
 
-    private static func mapActions(_ actions: [ToolbarActionState], windowUUID: UUID) -> [ToolbarElement] {
+    private static func mapActions(
+        _ actions: [ToolbarActionState],
+        state: ToolbarState,
+        windowUUID: UUID
+    ) -> [ToolbarElement] {
         return actions.map { action in
-            ToolbarElement(
+            let badgeImageName: String? = if state.isPrivateMode && action.canShowPrivateBrowsingBadge {
+                action.badgeImageName
+            } else {
+                if state.isLoginActionRequired && action.canShowWarningBadge {
+                    action.badgeImageName
+                } else {
+                    nil
+                }
+            }
+            return ToolbarElement(
                 iconName: action.iconName,
+                badgeImageName: badgeImageName,
```

Comment:
The `badgeImageName` should come from the action (and should be set through the middleware). It should not be calculated here.